### PR TITLE
Support `assume_role_arn` for database dynamic resources

### PR DESCRIPTION
**Pull request by greedy52**
Related PR:
- #12871 

Will update doc separately.

### CODE REVIEW

File: lib/config/fileconf.go [hunk 1]
```diff
@@ -1695,7 +1695,7 @@ type Databases struct {
 	// Databases is a list of databases proxied by the service.
 	Databases []*Database `yaml:"databases"`
 	// ResourceMatchers match cluster database resources.
-	ResourceMatchers []ResourceMatcher `yaml:"resources,omitempty"`
+	ResourceMatchers []DatabaseResourceMatcher `yaml:"resources,omitempty"`
```
Comment (smallinsky) [inline lib/config/fileconf.go:1698]: nit: 
Do you see any cons of adding the `AWS ResourceMatcherAWS` member to `ResourceMatcher` without adding a `DatabaseResourceMatcher`  ?. 
The `AWS` member in ResourceMatcher can be marked as optional and reused between kube app access in the future.

**Review by smallinsky — APPROVED** [review]

File: lib/config/fileconf.go (hunk 1 above)
Comment (greedy52) [reply lib/config/fileconf.go]: I originally added a new type to ensure that other resource matchers will throw error if `assume_role_arn` is specified. 

I removed the `DatabaseResourceMatcher` in https://github.com/gravitational/teleport/pull/28039/commits/4a9acaeb2bee4e5142d369943611afcca5491c77, and explicitly throwing errors for other two types.

**Review by avatus — APPROVED** [review]

File: lib/config/configuration.go [hunk 2]
```diff
@@ -1313,6 +1313,9 @@ func applyKubeConfig(fc *FileConfig, cfg *servicecfg.Config) error {
 	}
 
 	for _, matcher := range fc.Kube.ResourceMatchers {
+		if matcher.AWS.AssumeRoleARN != "" {
+			return trace.NotImplemented("assume_role_arn is not supported for kube resource matchers")
+		}
```
Comment (GavinFrazar) [inline lib/config/configuration.go:1318]: could we add a test to cover this in case we add support later

File: lib/config/configuration.go [hunk 3]
```diff
@@ -1512,6 +1519,9 @@ func applyAppsConfig(fc *FileConfig, cfg *servicecfg.Config) error {
 
 	// Configure resource watcher selectors if present.
 	for _, matcher := range fc.Apps.ResourceMatchers {
+		if matcher.AWS.AssumeRoleARN != "" {
+			return trace.NotImplemented("assume_role_arn is not supported for app resource matchers")
+		}
```
Comment (GavinFrazar) [inline lib/config/configuration.go:1524]: and a test here as well

File: lib/configurators/aws/aws.go [hunk 4]
```diff
@@ -1118,6 +1127,13 @@ func isTargetAWSAssumeRoleForDatabases(databases []*servicecfg.Database, targetA
 	})
 }
 
+func isTargetAWSAssumeRoleForResourceMatchers(resourceMatchers []services.ResourceMatcher, targetARN string) bool {
+	for _, resourceMatcher := range resourceMatchers {
+		return resourceMatcher.AWS.AssumeRoleARN == targetARN
```
Comment (GavinFrazar) [inline lib/configurators/aws/aws.go:1132]: this returns whether the first one matches

**Review by GavinFrazar — CHANGES_REQUESTED** [review]
looks good, just blocking on the configurator loop thing

File: lib/configurators/aws/aws.go (hunk 4 above)
Comment (greedy52) [reply lib/configurators/aws/aws.go:1132]: good catch!

**Review by GavinFrazar — APPROVED** [review]

**Pull request merged**